📖  (doc): Add a doc as a guidance to help users know how to consume the metrics and integrate it with other solutions (#1524)
operator-framework/operator-controller: docs/draft/howto/consuming-metrics.md

# Consuming Metrics

!!! warning
Metrics endpoints and ports are available as an alpha release and are subject to change in future versions.
The following procedure is provided as an example for testing purposes. Do not depend on alpha features in production clusters.

> In OLM v1, you can use the provided metrics with tools such as the [Prometheus Operator][prometheus-operator]. By default, Operator Controller and Catalog export metrics to the `/metrics` endpoint of each service.

Reviewer: ```suggestion
In OLM v1, you can use the provided metrics with tools such as the [Prometheus Operator][prometheus-operator]. By default, Operator Controller and catalogd export metrics to the `/metrics` endpoint of each service.
```

    image: curlimages/curl:latest
    command:
    - sh
    - -c
    - sleep 3600
    securityContext:
      runAsNonRoot: true
      readOnlyRootFilesystem: true
      runAsUser: 1000
      runAsGroup: 1000
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
    volumeMounts:
    - mountPath: /tmp/cert
      name: catalogd-cert
      readOnly: true
  volumes:
  - name: catalogd-cert
    secret:
      secretName: $OLM_SECRET
  securityContext:
    runAsNonRoot: true
  restartPolicy: Never
EOF
```

4. Access the pod:

```shell
kubectl exec -it curl-metrics-catalogd -n olmv1-system -- sh
```

5. Run the following command using the `TOKEN` value obtained above to check the metrics:

```shell
curl -v -k -H "Authorization: Bearer <TOKEN>" \
https://catalogd-service.olmv1-system.svc.cluster.local:7443/metrics
```

6. Run the following command to validate the certificates and token:
```shell
curl -v --cacert /tmp/cert/ca.crt --cert /tmp/cert/tls.crt --key /tmp/cert/tls.key \
-H "Authorization: Bearer <TOKEN>" \
https://catalogd-service.olmv1-system.svc.cluster.local:7443/metrics
```

---

## Integrating the metrics endpoints with third-party solutions

> In this case, you will need to provide usually, the certificates and the ServiceNames.
> To ilustrate those specific values required from OLMV1 see an example of the `ServiceMonitor`
> to scrape metrics for [Prometheus Operator][prometheus-operator]

Reviewer: ```suggestion
In many cases, you must provide the certificates and the `ServiceName` resources to integrate metrics endpoints with third-party solutions.
The following example illustrates how to create a `ServiceMonitor` resource to scrape metrics for the [Prometheus Operator][prometheus-operator] in OLM v1.
```

!!! note
    The following manifests are provided as a reference mainly to let you know how to configure the certificates. 
    The following procedure is not a complete guide to configuring the Prometheus Operator or how to integrate within.

>     Be aware that, for example, to integrate with [Prometheus Operator][prometheus-operator] you might need to adjust your 

Reviewer: ```suggestion
    To integrate with [Prometheus Operator][prometheus-operator] you might need to adjust your 
```
Minimalism nit: I think we can cut "be aware" and "for example" here.

echo $TOKEN
```

#### Deploy a Pod to Consume Metrics

Ensure that the Pod is deployed in a namespace labeled to enforce restricted permissions. Apply the following:

```shell
kubectl apply -f - <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: curl-metrics
  namespace: olmv1-system
spec:
  serviceAccountName: operator-controller-controller-manager
  containers:
  - name: curl
    image: curlimages/curl:latest
    command:
    - sh
    - -c
    - sleep 3600
    securityContext:
      runAsNonRoot: true
      readOnlyRootFilesystem: true
      runAsUser: 1000
      runAsGroup: 1000
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
    volumeMounts:
    - mountPath: /tmp/cert
      name: olm-cert
      readOnly: true
  volumes:
  - name: olm-cert
    secret:
      secretName: olmv1-cert
  securityContext:
    runAsNonRoot: true
  restartPolicy: Never
EOF
```

#### Access the Pod and Test Metrics

Access the pod:

```shell
kubectl exec -it curl-metrics -n olmv1-system -- sh
```

From the shell use the `TOKEN` value obtained above to check the metrics:

```shell
curl -v -k -H "Authorization: Bearer <TOKEN>" \

> https://operator-controller-controller-manager-metrics-service.olmv1-system.svc.cluster.local:8443/metrics

Author: ```suggestion
https://operator-controller-controller-manager-metrics-service.olmv1-system.svc.cluster.local:7443/metrics
```

> https://operator-controller-controller-manager-metrics-service.olmv1-system.svc.cluster.local:7443/metrics

Author: ```suggestion
https://operator-controller-service.olmv1-system.svc.cluster.local:8443/metrics
```

```

Validate using certificates and token:

```shell
curl -v --cacert /tmp/cert/ca.crt --cert /tmp/cert/tls.crt --key /tmp/cert/tls.key \
-H "Authorization: Bearer <TOKEN>" \

> https://operator-controller-controller-manager-metrics-service.olmv1-system.svc.cluster.local:7443/metrics

Author: ```suggestion
https://operator-controller-service.olmv1-system.svc.cluster.local:8443/metrics
```

> https://operator-controller-controller-manager-metrics-service.olmv1-system.svc.cluster.local:8443/metrics

Author: ```suggestion
https://operator-controller-controller-manager-metrics-service.olmv1-system.svc.cluster.local:7443/metrics
```

# Consuming Metrics

!!! warning
    Metrics endpoints and ports are available as an alpha release and are subject to change in future versions.
    The following procedure is provided as an example for testing purposes. Do not depend on alpha features in production clusters.

>     Additionally, any examples involving third-party solutions such as Prometheus are for reference only and are not officially supported.

Author: ```suggestion
```

echo $TOKEN
```

#### Deploy a Pod to Consume Metrics

Ensure that the Pod is deployed in a namespace labeled to enforce restricted permissions. Apply the following:

```shell
kubectl apply -f - <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: curl-metrics
  namespace: olmv1-system
spec:
  serviceAccountName: operator-controller-controller-manager
  containers:
  - name: curl
    image: curlimages/curl:latest
    command:
    - sh
    - -c
    - sleep 3600
    securityContext:
      runAsNonRoot: true
      readOnlyRootFilesystem: true
      runAsUser: 1000
      runAsGroup: 1000
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
    volumeMounts:
    - mountPath: /tmp/cert
      name: olm-cert
      readOnly: true
  volumes:
  - name: olm-cert
    secret:
      secretName: olmv1-cert
  securityContext:
    runAsNonRoot: true
  restartPolicy: Never
EOF
```

#### Access the Pod and Test Metrics

Access the pod:

```shell
kubectl exec -it curl-metrics -n olmv1-system -- sh
```

From the shell use the `TOKEN` value obtained above to check the metrics:

```shell
curl -v -k -H "Authorization: Bearer <TOKEN>" \

> https://operator-controller-service.olmv1-system.svc.cluster.local:8443/metrics

Author: @tylerslaton THANK You for check that it has something wrong
The name of the service was wrong when I copied and pasted the command.
It is fine now as you can check in the description

c/c @LalatenduMohanty

> https://operator-controller-controller-manager-metrics-service.olmv1-system.svc.cluster.local:8443/metrics

Reviewer 2: Following the tutorial with a fresh cluster, this curl command fails:
`curl: (6) Could not resolve host: operator-controller-controller-manager-metrics-service.olmv1-system.svc.cluster.local`

Reviewer 2: are you missing a step where we need to create a service or something? or did I miss something? I'm just using the default "make run" target to spin up a cluster.

Author: Oh the port changed to 7443
I will update here good catcher

Validate using certificates and token:

```shell
curl -v --cacert /tmp/cert/ca.crt --cert /tmp/cert/tls.crt --key /tmp/cert/tls.key \
-H "Authorization: Bearer <TOKEN>" \
https://operator-controller-service.olmv1-system.svc.cluster.local:8443/metrics
```

---

## CatalogD Metrics

### Step 1: Enable Access

To enable access to the CatalogD metrics, create a `ClusterRoleBinding` for the CatalogD service account:

```shell
kubectl create clusterrolebinding catalogd-metrics-binding \
   --clusterrole=catalogd-metrics-reader \
   --serviceaccount=olmv1-system:catalogd-controller-manager
```

### Step 2: Validate Access Manually

#### Create a Token and Extract Certificates

Generate a token and get the required certificates:

```shell
TOKEN=$(kubectl create token catalogd-controller-manager -n olmv1-system)
echo $TOKEN
```

#### Deploy a Pod to Consume Metrics

From the shell use the `TOKEN` value obtained above to check the metrics:

```shell
OLM_SECRET=$(kubectl get secret -n olmv1-system -o jsonpath="{.items[*].metadata.name}" | tr ' ' '\n' | grep '^catalogd-service-cert')

> echo $OLM_SECRET

Author: @tylerslaton @LalatenduMohanty

Here as well was not working now it is fine

```
$ OLM_SECRET=$(kubectl get secret -n olmv1-system -o jsonpath="{.items[*].metadata.name}" | tr ' ' '\n' | grep '^catalogd-service-cert')
camilam@Camilas-MacBook-Pro ~/go/src/github/operator-framework/operator-controller (remove-monitor-prometheues) $ echo $OLM_SECRET
catalogd-service-cert-v1.2.0-rc4-19-g099a6cf

```

to ensure that metrics are properly scraped. This will depend on how Prometheus is configured and, for example, the namespace  
where the `ServiceMonitor` is applied.

### For Operator-Controller

```shell
kubectl apply -f - <<EOF
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    control-plane: operator-controller-controller-manager
  name: controller-manager-metrics-monitor
  namespace: system
spec:
  endpoints:
    - path: /metrics
      port: https
      scheme: https
      bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
      tlsConfig:
        insecureSkipVerify: false 
        serverName: operator-controller-controller-manager-metrics-service.olmv1-system.svc
        ca:
          secret:
            name: olmv1-cert
            key: ca.crt
        cert:
          secret:
            name: olmv1-cert
            key: tls.crt
        keySecret:
          name: olmv1-cert
          key: tls.key
  selector:
    matchLabels:
      control-plane: operator-controller-controller-manager
EOF
```

### For CatalogD

```shell

> OLM_SECRET=$(kubectl get secret -n olmv1-system -o jsonpath="{.items[?(@.metadata.name | startswith('catalogd-service-cert'))].metadata.name}")

Author: @tylerslaton @LalatenduMohanty

Here as well was not working now it is fine

```
$ OLM_SECRET=$(kubectl get secret -n olmv1-system -o jsonpath="{.items[*].metadata.name}" | tr ' ' '\n' | grep '^catalogd-service-cert')
camilam@Camilas-MacBook-Pro ~/go/src/github/operator-framework/operator-controller (remove-monitor-prometheues) $ echo $OLM_SECRET
catalogd-service-cert-v1.2.0-rc4-19-g099a6cf

```

    securityContext:
      runAsNonRoot: true
      readOnlyRootFilesystem: true
      runAsUser: 1000
      runAsGroup: 1000
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
    volumeMounts:
    - mountPath: /tmp/cert
      name: catalogd-cert
      readOnly: true
  volumes:
  - name: catalogd-cert
    secret:
      secretName: $OLM_SECRET
  securityContext:
    runAsNonRoot: true
  restartPolicy: Never
EOF
```

#### Access the Pod and Test Metrics

Access the pod:

```shell
kubectl exec -it curl-metrics-catalogd -n olmv1-system -- sh
```

From the shell use the `TOKEN` value obtained above to check the metrics:

```shell
curl -v -k -H "Authorization: Bearer <TOKEN>" \
https://catalogd-service.olmv1-system.svc.cluster.local:7443/metrics
```

Validate using certificates and token:

```shell
curl -v --cacert /tmp/cert/ca.crt --cert /tmp/cert/tls.crt --key /tmp/cert/tls.key \
-H "Authorization: Bearer <TOKEN>" \
https://catalogd-service.olmv1-system.svc.cluster.local:7443/metrics
```

---

## Enabling Integration with Prometheus

If using [Prometheus Operator][prometheus-operator], create a `ServiceMonitor` to scrape metrics:

!!! note

> The following manifests are provided as examples. You may need to configure certain settings, such as `serviceMonitorSelector`,  
> to ensure that metrics are properly scraped. This will depend on how Prometheus is configured and, for example, the namespace  
> where the `ServiceMonitor` is applied.

Reviewer: ```suggestion
The following manifests are provided as examples. The following procedure is not a complete guide to configuring the Prometheus Operator. 
You might need to adjust your configuration settings, such as the `serviceMonitorSelector` resource, and the namespace where you apply the `ServiceMonitor` resource to ensure that metrics are properly scraped.
```

> ### For Operator-Controller

Reviewer: ```suggestion
1. Apply the following YAML to enable metrics scraping for Operator Controller: 
```

```shell
kubectl apply -f - <<EOF
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    control-plane: operator-controller-controller-manager
  name: controller-manager-metrics-monitor
  namespace: olmv1-system
spec:
  endpoints:
    - path: /metrics
      port: https
      scheme: https
      bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
      tlsConfig:
        insecureSkipVerify: false 
        serverName: operator-controller-service.olmv1-system.svc
        ca:
          secret:
            name: olmv1-cert
            key: ca.crt
        cert:
          secret:
            name: olmv1-cert
            key: tls.crt
        keySecret:
          name: olmv1-cert
          key: tls.key
  selector:
    matchLabels:
      control-plane: operator-controller-controller-manager
EOF
```

> ### For CatalogD

Reviewer: ```suggestion
2. Run the following command to ? 
```

```shell
OLM_SECRET=$(kubectl get secret -n olmv1-system -o jsonpath="{.items[*].metadata.name}" | tr ' ' '\n' | grep '^catalogd-service-cert')
echo $OLM_SECRET
```

> 

Reviewer: 3. Apply the following YAML to enable metrics scraping for catalogd:

```shell
kubectl apply -f - <<EOF
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    control-plane: catalogd-controller-manager
  name: catalogd-metrics-monitor
  namespace: olmv1-system
spec:
  endpoints:
    - path: /metrics
      port: https
      scheme: https
      bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
      tlsConfig:
        serverName: catalogd-service.olmv1-system.svc
        insecureSkipVerify: false
        ca:
          secret:
            name: $OLM_SECRET
            key: ca.crt
        cert:
          secret:
            name: $OLM_SECRET
            key: tls.crt
        keySecret:
          name: $OLM_SECRET
          key: tls.key
  selector:
    matchLabels:
      control-plane: catalogd-controller-manager
EOF
```

> 

Reviewer: Is there a simple verification step we can add to help the user know that Prometheus Operator is running?
